<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=description content="CCBR whoLe genOme-sequencinG Analysis pipeliNe (LOGAN)"><meta name=author content="CCR Collaborative Bioinformatics Resource"><link href=https://ccbr.github.io/LOGAN/dev/ rel=canonical><link rel="shortcut icon" href=img/favicon.ico><title>CCBR LOGAN</title><link rel=stylesheet href=css/theme.css><link rel=stylesheet href=css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><link href=css/custom.css rel=stylesheet><link href=css/version-select.css rel=stylesheet><script>
        // Current page data
        var mkdocs_page_name = "Home";
        var mkdocs_page_input_path = "index.md";
        var mkdocs_page_url = "/LOGAN/dev/";
      </script><!--[if lt IE 9]>
      <script src="js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=. class="icon icon-home"> CCBR LOGAN </a><div role=search> <form id=rtd-search-form class=wy-form action=./search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul class=current> <li class="toctree-l1 current"><a class="reference internal current" href=#>Home</a> <ul class=current> <li class=toctree-l2><a class="reference internal" href=#overview>Overview</a> </li> <li class=toctree-l2><a class="reference internal" href=#dependencies>Dependencies</a> </li> <li class=toctree-l2><a class="reference internal" href=#set-up>Set up</a> <ul> <li class=toctree-l3><a class="reference internal" href=#biowulf>Biowulf</a> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#usage>Usage</a> <ul> <li class=toctree-l3><a class="reference internal" href=#input-files>Input Files</a> </li> <li class=toctree-l3><a class="reference internal" href=#genome>Genome</a> <ul> <li class=toctree-l4><a class="reference internal" href=#hg38-has-options-for-either>hg38 has options for either</a> </li> </ul> </li> <li class=toctree-l3><a class="reference internal" href=#operating-modes>Operating Modes</a> <ul> <li class=toctree-l4><a class="reference internal" href=#1-paired-tumornormal-mode>1. Paired Tumor/Normal Mode</a> </li> <li class=toctree-l4><a class="reference internal" href=#2-tumor-only-mode>2. Tumor only mode</a> </li> <li class=toctree-l4><a class="reference internal" href=#calling-mode>Calling Mode</a> </li> <li class=toctree-l4><a class="reference internal" href=#optional-arguments>Optional Arguments</a> </li> </ul> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#running-logan>Running LOGAN</a> <ul> <li class=toctree-l3><a class="reference internal" href=#pipeline-tools-and-overview>Pipeline Tools and Overview</a> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#contribute>Contribute</a> </li> <li class=toctree-l2><a class="reference internal" href=#references>References</a> </li> </ul> </li> </ul> <p class=caption><span class=caption-text>Pipeline</span></p> <ul> <li class=toctree-l1><a href=user-guide/pipeline/ class="reference internal">Tumor-Normal</a> </li> </ul> <p class=caption><span class=caption-text>Tools</span></p> <ul> <li class=toctree-l1><a href=user-guide/tool_comparisons/ class="reference internal">Comparisons</a> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=.>CCBR LOGAN</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=. class="icon icon-home" aria-label=Docs></a></li> <li class="breadcrumb-item active">Home</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/CCBR/LOGAN/edit/main/index.md>Edit on CCBR/LOGAN</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <h1 id=logan>LOGAN ðŸ”¬</h1> <p><strong><em>whoLe genOme-sequencinG Analysis pipeliNe</em></strong></p> <p><a href=https://github.com/CCBR/LOGAN/actions/workflows/build.yml><img alt=build src=https://github.com/CCBR/LOGAN/actions/workflows/build.yml/badge.svg></a> <a href=https://doi.org/10.5281/zenodo.14907169><img alt=DOI src=https://zenodo.org/badge/DOI/10.5281/zenodo.14907169.svg></a> <a href=https://hub.docker.com/r/nciccbr/ccbr_wes_base><img alt="Docker Pulls" src=https://img.shields.io/docker/pulls/nciccbr/ccbr_wes_base></a> <a href=https://github.com/ccbr/LOGAN/issues><img alt=issues src="https://img.shields.io/github/issues/ccbr/LOGAN?color=brightgreen"></a> <a href=https://github.com/ccbr/LOGAN/blob/master/LICENSE><img alt=license src=https://img.shields.io/github/license/ccbr/LOGAN></a> <a href=https://github.com/ccbr/LOGAN/releases/latest><img alt=release src="https://img.shields.io/github/v/release/CCBR/LOGAN?color=blue&label=latest%20release"></a></p> <p>Call germline and somatic variants, CNVs, and SVs and annotate variants!</p> <h2 id=overview>Overview</h2> <p>Welcome to LOGAN! Before getting started, we highly recommend reading through <a href=https://ccbr.github.io/LOGAN>LOGAN's documentation</a>.</p> <p>LOGAN is a comprehensive whole genome-sequencing pipeline following the Broad's set of best practices. It relies on technologies like <a href=https://singularity.lbl.gov/ >Singularity<sup>1</sup></a> to maintain the highest-level of reproducibility. The pipeline consists of a series of data processing and quality-control steps orchestrated by <a href=https://nextflow.io/ >Nextflow<sup>2</sup></a>, a flexible and scalable workflow management system, to submit jobs to a cluster or cloud provider.</p> <p>Before getting started, we highly recommend reading through the <a href=https://ccbr.github.io/LOGAN/usage/run/ >usage</a> section of each available sub command.</p> <p>For more information about issues or trouble-shooting a problem, please checkout our <a href=faq/questions.md>FAQ</a> prior to <a href=https://github.com/ccbr/LOGAN/issues>opening an issue on Github</a>.</p> <p>Original pipelining and code forked from the CCBR Exome-seek Pipeline <a href=https://github.com/CCBR/XAVIER>Exome-seek</a> and <a href=https://github.com/openOmics/genome-seek>OpenOmics</a></p> <h2 id=dependencies>Dependencies</h2> <p><strong>Requires:</strong> <code>singularity&gt;=3.5</code> <code>nextflow&gt;=22.10.2</code></p> <p><a href=https://singularity.lbl.gov/all-releases>singularity</a> must be installed on the target system. Nextflow orchestrates the execution of each step in the pipeline. To guarantee the highest level of reproducibility, each step relies on versioned images from <a href=https://hub.docker.com/orgs/nciccbr/repositories>DockerHub</a>. Nextflow uses singularity to pull these images onto the local filesystem prior to job execution, and as so, nextflow and singularity are the only two dependencies.</p> <h2 id=set-up>Set up</h2> <p>LOGAN is installed on the <a href=#biowulf>Biowulf HPC</a>. For installation in other execution environments, refer to the <a href=https://ccbr.github.io/LOGAN/ >docs</a>.</p> <h3 id=biowulf>Biowulf</h3> <p>LOGAN is available on <a href=https://hpc.nih.gov/ >Biowulf</a> in the <code>ccbrpipeliner</code> module. You'll first need to start an interactive session, then load the module:</p> <pre><code class=language-sh># start an interactive node
sinteractive --mem=2g --cpus-per-task=2 --gres=lscratch:200

# load the ccbrpipeliner module
module load ccbrpipeliner
</code></pre> <h2 id=usage>Usage</h2> <h3 id=input-files>Input Files</h3> <p>LOGAN supports inputs of either</p> <ol> <li>paired end fastq files</li> </ol> <p><code>--fastq_input</code>- A glob can be used to include all FASTQ files. Like <code>--fastq_input "*R{1,2}.fastq.gz"</code>. Globbing requires quotes.</p> <ol> <li>Pre aligned BAM files with BAI indices</li> </ol> <p><code>--bam_input</code>- A glob can be used to include all FASTQ files. Like <code>--bam_input "*.bam"</code>. Globbing requires quotes.</p> <ol> <li>A sheet that indicates the sample name and either FASTQs or BAM file locations</li> </ol> <p><code>--fastq_file_input</code>- A headerless tab delimited sheet that has the sample name, R1, and R2 file locations</p> <p>Example</p> <pre><code class=language-bash>Sample1_TUMOR  /path/to/the/fastq/folder/Sample1_TUMOR.R1.fq.gz /path/to/the/fastq/folder/Sample1_TUMOR.R2.fq.gz
Sample1_NORMAL  /path/to/the/fastq/folder/Sample1_NORMAL.R1.fq.gz /path/to/the/fastq/folder/Sample1_NORMAL.R2.fq.gz
Sample2_TUMOR  /path/to/the/fastq/folder/Sample2_TUMOR.R1.fq.gz /path/to/the/fastq/folder/Sample2_TUMOR.R2.fq.gz
Sample2_NORMAL  /path/to/the/fastq/folder/Sample2_NORMAL.R1.fq.gz /path/to/the/fastq/folder/Sample2_NORMAL.R2.fq.gz
</code></pre> <p><code>--bam_file_input</code> - A headerless tab delimited sheet that has the sample name, bam, and bam index (bai) file locations</p> <p>Example</p> <pre><code class=language-bash>Sample1_TUMOR   /path/to/the/BAM/folder/Sample1_TUMOR.bam  /path/to/the/BAM/folder/Sample1_TUMOR.bam.bai
Sample1_NORMAL  /path/to/the/BAM/folder/Sample1_NORMAL.bam  /path/to/the/BAM/folder/Sample1_NORMAL.bam.bai
</code></pre> <h3 id=genome>Genome</h3> <p><code>--genome</code> - A flag to indicate which genome to run. hg38, hg19 and mm10 are supported.<br> Example: <code>--genome hg38</code> to run the hg38 genome</p> <p><code>--genome hg19</code> and <code>--genome mm10</code> are also supported</p> <h4 id=hg38-has-options-for-either>hg38 has options for either</h4> <p><code>--genome hg38</code> - Based off the GRCh38.d1.vd1.fa which is consistent with TCGA/GDC processing pipelines</p> <p><code>--genome hg38_sf</code> - Based off the Homo_sapiens_assembly38.fasta which is derived from the Broad Institute/NCI Sequencing Facility The biggest difference between the two is that GRCh38.d1.vd1.fa only the GCA_000001405.15_GRCh38_no_alt_analysis_set, Sequence Decoys (GenBank Accession GCA_000786075), and Virus Sequences. Homo_sapiens_assembly38.fasta has HLA specific contigs which may not be compatible with certain downstream tools.</p> <h3 id=operating-modes>Operating Modes</h3> <h4 id=1-paired-tumornormal-mode>1. Paired Tumor/Normal Mode</h4> <p>Required for Paired Tumor/Normal Mode</p> <p><code>--sample_sheet</code> In Paired mode a sample sheet must be provided with the basename of the Tumor and Normal samples. This sheet must be Tab separated with a header for Tumor and Normal.</p> <p>Example</p> <pre><code class=language-bash>Tumor  Normal
Sample1_TUMOR  Sample1_Normal
Sample2_TUMOR  Sample1_Normal
</code></pre> <h4 id=2-tumor-only-mode>2. Tumor only mode</h4> <p>No additional flags for sample sheet are required as all samples will be used to call variants</p> <h4 id=calling-mode>Calling Mode</h4> <p>Adding flags determines SNV (germline and/or somatic), SV, and/or CNV calling modes</p> <p><code>--vc</code> or <code>--snv</code> - Enables somatic SNV calling using mutect2, vardict, varscan, octopus, deepsomatic, strelka (TN only), MUSE (TN only), and lofreq (TN only)</p> <p><code>--gl</code> or <code>--germline</code> - Enables germline calling using Deepvariant</p> <p><code>--sv</code> or <code>--structural</code>- Enables somatic SV calling using Manta, GRIDSS, and SVABA</p> <p><code>--cnv</code> or <code>--copynumber</code>- Enables somatic CNV calling using FREEC, Sequenza, ASCAT, CNVKit, and Purple (hg19/hg38 only)</p> <h4 id=optional-arguments>Optional Arguments</h4> <p><code>--callers</code> - Comma separated argument for selecting only specified callers, the default is to use all. Example: <code>--callers mutect2,octopus</code></p> <p><code>--cnvcallers</code> - Comma separated argument for selecting only specified CNV callers, the default is to use all. Example: <code>--cnvcallers purple</code></p> <p><code>--svcallers</code> - Comma separated argument for selecting only specified SV callers, the default is to use all. Example: <code>--svcallers gridss</code></p> <p><code>--ffpe</code> - Adds additional filtering for FFPE by detecting strand orientation bias using SOBDetector.</p> <p><code>--exome</code> - When using exome data, this flag limits calling to intervals provided in target bed to reduce time and to account for exome sequencing specific parameters. An intervals file is required.</p> <p><code>--indelrealign</code> - Enables indel realignment using the GATK pipeline when running alignment steps. May be helpful for certain callers (VarScan, VarDict) that do not have local haplotype reassembly.</p> <h2 id=running-logan>Running LOGAN</h2> <p>Example of Tumor_Normal calling mode</p> <pre><code class=language-bash># Step 0: Set up

sinteractive --mem=8g -N 1 -n 4
module load ccbrpipeliner # v8 

# set up directories

DATADIR=&quot;/path/to/fastq/folder&quot;

# Step1: Initilalization

logan init \
--output logan_output

# Step2: stub-run

logan run  \
--output logan_output \
--mode local -profile ci_stub \
--genome hg38 \
--sample_sheet samplesheet.txt \
--fastq_file_input logan_fastq_input.txt \
-stub --vc --sv --cnv

# Step3: Full run
logan run  \
--output logan_output \
--mode slurm -profile slurm \
--genome hg38 \
--sample_sheet samplesheet.txt \
--fastq_file_input logan_fastq_input.txt \
--vc --sv --cnv

# NOTE: In case of globbing instead of using a fastq input sheet

logan run  \
--output logan_output \
--mode local -profile ci_stub \
--genome hg38 \
--sample_sheet samplesheet.txt \
--fastq_input &quot;/path/to/fastq/folder/*R{1,2}.fastq.gz&quot; \
-stub --vc --sv --cnv

logan run  \
--output logan_output \
--mode slurm -profile slurm \
--genome hg38 \
--sample_sheet samplesheet.txt \
--fastq_input &quot;/path/to/fastq/folder/*R{1,2}.fastq.gz&quot; \
--vc --sv --cnv

</code></pre> <p>Example of Tumor only calling mode</p> <pre><code class=language-bash># preview the logan jobs that will run
logan run --mode local -profile ci_stub --genome hg38 --outdir logan_output --fastq_input &quot;*R{1,2}.fastq.gz&quot; --callers octopus,mutect2 -preview --vc --sv --cnv
# run a stub/dryrun of the logan jobs
logan run --mode local -profile ci_stub --genome hg38 --outdir logan_output --fastq_input &quot;*R{1,2}.fastq.gz&quot; --callers octopus,mutect2 -stub --vc --sv --cnv
# launch a logan run on slurm with the test dataset
logan run --mode slurm -profile slurm --genome hg38 --outdir logan_output --fastq_input &quot;*R{1,2}.fastq.gz&quot; --callers octopus,mutect2 --vc --sv --cnv
</code></pre> <h3 id=pipeline-tools-and-overview>Pipeline Tools and Overview</h3> <p><img alt="alt text" src=docs/LOGAN.png></p> <h2 id=contribute>Contribute</h2> <p>This site is a living document, created for and by members like you. LOGAN is maintained by the members of CCBR and is improved by continuous feedback! We encourage you to contribute new content and make improvements to existing content via pull request to our <a href=https://github.com/ccbr/LOGAN/pulls>repository</a>.</p> <h2 id=references>References</h2> <p>This repo was originally generated from the <a href=https://github.com/CCBR/CCBR_NextflowTemplate>CCBR Nextflow Template</a>.</p> <p><sup><strong>1.</strong> Kurtzer GM, Sochat V, Bauer MW (2017). Singularity: Scientific containers for mobility of compute. PLoS ONE 12(5): e0177459.</sup></p> </div> </div><script async src="https://www.googletagmanager.com/gtag/js?id=G-D3SL9V30KL"></script> <script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-D3SL9V30KL'); // Replace with your tracking ID
</script> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/CCBR/LOGAN class="fa fa-code-fork" style="color: #fcfcfc"> CCBR/LOGAN</a> </span> <span><a href=user-guide/pipeline/ style="color: #fcfcfc">Next &raquo;</a></span> </span> </div> <script src=js/jquery-3.6.0.min.js></script> <script>var base_url = ".";</script> <script src=js/theme_extra.js></script> <script src=js/theme.js></script> <script src=js/custom.js></script> <script src=search/main.js></script> <script src=js/version-select.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> <!--
MkDocs version : 1.6.1
Build Date UTC : 2025-09-10 17:24:45.430736+00:00
--> 